### ISSUE TITLE
`buildctl dial-stdio` has exited with exit status 137

### ISSUE THREAD

**Issue by jedevc** [dagger/dagger#8329]
Occasionally, this error appears:

```
Error: response from query: Post "http://:mem/query": command [docker exec -i dagger-engine-d522a778c2389390 buildctl dial-stdio] has exited with exit status 137, make sure the URL is valid, and Docker 18.09 or later is installed on the remote host: stderr=
```

Seen in:
- https://discord.com/channels/707636530424053791/1003718839739105300/1280688927652450399 from @shykes 
- https://discord.com/channels/707636530424053791/1120503349599543376/1272596156848934973 by @sagikazarmark 
- https://discord.com/channels/707636530424053791/1263881471026925688/1268233582111887450

The `http://:mem/query` is very odd, but it seems to come from: https://github.com/dagger/dagger/blob/7dd8635b56a24b60a66b89f44c6aeedd48853ba8/engine/client/client.go#L1144

Not entirely sure why it's set to that (blame points to the git blame hole that is https://github.com/dagger/dagger/pull/5757).

**Comment by idlsoft** [comment]
I'm getting this error consistently when I run on a Mac, but not on Linux.
Dagger 0.14.0.
Running:
dagger shell -q s1.dag

s1.dag:
```
#!/usr/bin/env dagger shell -q
ctr1=$(.container | from alpine | with-exec echo,hello,world)
..echo $ctr1 | stdout
```

Linux (works):
```
docker version
Client: Docker Engine - Community
 Version:           27.3.1
 API version:       1.47
 Go version:        go1.22.7
 Git commit:        ce12230
 Built:             Fri Sep 20 11:41:03 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          27.3.1
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.22.7
  Git commit:       41ca978
  Built:            Fri Sep 20 11:41:03 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.23
  GitCommit:        57f17b0a6295a39009d861b89e3b3b87b005ca27
 runc:
  Version:          1.1.14
  GitCommit:        v1.1.14-0-g2c9f560
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Mac (doesn't work, using Colima):
```
docker version
Client: Docker Engine - Community
 Version:           27.3.1
 API version:       1.47
 Go version:        go1.23.1
 Git commit:        ce1223035a
 Built:             Fri Sep 20 11:01:47 2024
 OS/Arch:           darwin/arm64
 Context:           colima

Server: Docker Engine - Community
 Engine:
  Version:          27.3.1
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.22.7
  Git commit:       41ca978
  Built:            Fri Sep 20 11:41:54 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.7.22
  GitCommit:        7f7fdf5fed64eb6a7caf99b3e12efcf9d60e311c
 runc:
  Version:          1.1.14
  GitCommit:        v1.1.14-0-g2c9f560
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by stavros-k** [comment]
Just to add some observations here. (See more details here https://discord.com/channels/707636530424053791/1297342703436038154/1297689740233015307)

I had similar issue where dial-stdio was exiting with `1`, but it turned out to be due to CPU/Memory load.
ie, I had jenkins run multiple `dagger run go run ...` on the same host. Once I tuned the number of executors, I don't see the issue anymore

**Comment by helderco** [comment]
I'm getting a similar error, but not with `":mem:"`:

```
Error: response from query: Post "http://dagger/query": command [docker exec -i dagger-engine-v0.14.0 buildctl dial-stdio] has exited with exit status 137, make sure the URL is valid, and Docker 18.09 or later is installed on the remote host: stderr=
```

**Comment by jedevc** [comment]
@helderco I replaced the `:mem:` with `dagger` in https://github.com/dagger/dagger/pull/8463

But yeah, still not fully sure what the root cause is.

**Comment by cwlbraa** [comment]
poking corroborates @stavros-k's suspicions: this is the engine getting oomkilled. it also explains the :mem: thing, which is a weird bit of output, but at least gives you a hint to boost your container runtime VM's memory allocation.

this repros consistently running hack dev on Colima with default args. it also repros with a customized bigger vm with the same default memory allocation of 2gb (eg `colima start --vm-type=vz --vz-rosetta --cpu=4 --memory=2 --disk=100`).

it doesn't repro with 16gb of memory, regardless of other parameters, nor does it repro with 4gb and 4cpus.

**Comment by jedevc** [comment]
Ahhhh, that kinda does explain the `:mem:` thing.

Kinda interesting to me that we get OOM with only 2GB of memory - are we really *that* bad with memory usage? I'd be curious to know what all our memory is actually doing, while we're not the *lightest* piece of software out there, I am surprised that we'd consistently be using that much - I'd expect our subprocesses to be the more demanding ones.

Does this happen right in the middle of a build? Or during the connection stage?

Out of curiosity, does the docker daemon in colima keep running happily? Can we get logs from the crashed container? If so, I think a good error behavior in this case would be to `docker logs` the engine container, and give those to the user.

**Comment by stavros-k** [comment]
In my case it was during (multiple parallel) npm build which is known for eating ram.

Besides some clearer error message, I'm not sure what else can be done at this time.

Maybe in the far far (did I say far?) future, dagger can schedule tasks with queues when it sees a memory pressure.

**Comment by cwlbraa** [comment]
> Does this happen right in the middle of a build? Or during the connection stage?

the crashes are definitely in the middle, hard to tell under our tests the difference between the connection stage and later.

> Out of curiosity, does the docker daemon in colima keep running happily? Can we get logs from the crashed container? If so, I think a good error behavior in this case would be to docker logs the engine container, and give those to the user.

docker daemon keeps running happily, yes, but there's nothing helpful in the logs.

**Comment by EmmS21** [comment]
I'm running a pipeline to extract individual questions from exam papers.

I queue up PDFs from a list in a database, download each from Google Drive and extract the text & images. I send this to AI to extract and enrich individual questions with some metadata.

This is a queue of ~2000 PDFs ranging in size from 4 to 15 pages.

I keep encountering this error:

```
17:23:55 WRN rotate dagger.cloud token for full url
Error: Post "http://dagger/query": command [docker exec -i dagger-engine-v0.17.0-llm.2 buildctl dial-stdio] has exited with exit status 137, make sure the URL is valid, and Docker 18.09 or later is installed on the remote host: stderr=
```

The pipeline runs through a batch of PDFs (~50-80) and then fails with this error. This has happened multiple times.

The only other logs I see are logs I added to track whether each step was being successfully completed for each PDF:

```
DEBUG: Creating container with file of size 3585 bytes
DEBUG: Writing extraction result to file in container...
DEBUG: JSON contains 3 questions
DEBUG: Raw insert_result_json: '{"success": true, "extraction_id": "67d30d84ea4ae16b7f91c4f7", "total_questions": 3}\n'
SUCCESS: Inserted 3 questions for document 67c3b79e17d230a3beca20d9
SUCCESS: Document 67c3b79e17d230a3beca20d9 marked as processed
Processing document 67c3b79e17d230a3beca20da, file ID: 1e4rcZAZ4Mczeh5AJ-xWPPxyiShs2M5r9
DEBUG: Download result: {"success": true, "file_id": "1e4rcZAZ4Mczeh5AJ-xWPPxyiShs2M5r9", "file_name": "Paper-3:33Oct.pdf", "pdf_path": "/app/Paper-3:33Oct.pdf.pdf"}
...
DEBUG: Extraction result type: <class 'str'>, length: 4487
DEBUG: First 100 chars of extraction result: {
  "success": true,
  "file_id": "1e4rcZAZ4Mczeh5AJ-xWPPxyiShs2M5r9",
  "file_name": "Paper-3:33Oct
DEBUG: Last 100 chars of extraction result: ch year. Justify your answer",
          "marks": 5
        }
      ],
      "marks": 25
    }
  ]
}
DEBUG: Successfully parsed extraction JSON with 3 questions
DEBUG: JSON structure keys: ['success', 'file_id', 'file_name', 'collection_name', 'total_questions', 'extracted_questions']
DEBUG: Creating container with file of size 4487 bytes
DEBUG: Writing extraction result to file in container...
DEBUG: JSON contains 3 questions
```

Assuming this is a memory issue?

**Comment by cwlbraa** [comment]
@EmmS21, yes, smells like a memory issue to me. You can confirm that by seeing if it gets further with a larger memory allocation or via watching the engine eat up memory with something like `atop`.
